### PR TITLE
HDDS-4143. Implement a version factory for OM Apply Transaction that …

### PR DESCRIPTION
**Pull request by avijayanhwx**
…uses the implementation based on layout version.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OmVersionFactory.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+      String type, OMLayoutFeature layoutFeature,
+      Function<OMRequest, OMClientRequest> function) {
+    VersionFactoryKey key = new VersionFactoryKey.Builder()
+        .key(type).version(layoutFeature.layoutVersion()).build();
+    writeRequestFactory.register(key, function);
+  }
+
+  public OMClientRequest getClientRequest(OMRequest omRequest) {
+    Function<OMRequest, ? extends OMClientRequest> instance =
+        writeRequestFactory.getInstance(getVersionFactoryKey(omRequest));
```
Comment (linyiqun) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OmVersionFactory.java:118]: When input omRequest type is not existed, we should return IllegalStateException error as original logic OzoneManagerRatisUtils#createClientRequest did.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 2]
```diff
@@ -84,86 +61,26 @@
...
   private OzoneManagerRatisUtils() {
   }
-  /**
-   * Create OMClientRequest which encapsulates the OMRequest.
-   * @param omRequest
-   * @return OMClientRequest
-   * @throws IOException
-   */
-  public static OMClientRequest createClientRequest(OMRequest omRequest) {
-    Type cmdType = omRequest.getCmdType();
```
Comment (linyiqun) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:94]: Can we keep OzoneManagerRatisUtils#createClientRequest unchanged and invoke omVersionFactory#getClientRequest Here?
Now OzoneManagerRatisUtils defines the request/response functions, I prefer to keep this function there. This is just my personal taste for this change.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/upgrade/TestOmVersionFactory.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMRequest;
+import org.apache.hadoop.ozone.upgrade.VersionFactoryKey;
+import org.apache.hadoop.ozone.upgrade.VersionedInstanceSupplierFactory;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * Test OmVersionFactory.
+ */
+public class TestOmVersionFactory {
```
Comment (linyiqun) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/upgrade/TestOmVersionFactory.java:39]: Can we add a test case that using existed OM request and verify its layout version is set as expected?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java [hunk 4]
```diff
@@ -290,6 +290,11 @@ public void validateRequest(OMRequest omRequest) throws OMException {
       throw new OMException("ClientId is null",
           OMException.ResultCodes.INVALID_REQUEST);
     }
+
+    if (omRequest.getLayoutVersion() == null) {
```
Comment (linyiqun) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:294]: We might add a comment to say why we should pass layout version in write type request. That will be good understanding.

**Review by linyiqun — COMMENTED** [review]
Hi @avijayanhwx , thanks for attaching the PR.
Overall looks good, just some my minor comments.

**Comment by avijayanhwx** [comment]
Thanks for the comments @linyiqun. I had closed the PR since I am changing the registration logic of new OM client requests, to an automatic registration based on Annotation. I will keep your comments in mind in my next PR.

**Pull request closed without merging**